File: scripts/tmux.py
# /// script
# dependencies = [
#   "dataclasses",
#   "dataclasses-json",
#   "invoke",
# ]
# ///
from dataclasses import dataclass
import os
import sys
from time import sleep
import invoke


from dataclasses import dataclass, field, fields
from dataclasses_json import dataclass_json, DataClassJsonMixin
from pathlib import Path
from typing import ClassVar

@dataclass
class Window(DataClassJsonMixin):
    delimeter: ClassVar[str] = "\t"
    session_name: str = field(default="", metadata={"fmt": "#{session_name}"})
    window_index: int = field(default=0, metadata={"fmt": "#{window_index}"})
    window_name: str = field(default="", metadata={
                             "fmt": ":#{window_name}"})
    window_active: bool = field(default=False, metadata={
                                "fmt": "#{window_active}"})
    window_flags: str = field(default="", metadata={"fmt": ":#{window_flags}"})
    window_layout: str = field(default="", metadata={
                               "fmt": "#{window_layout}"})

    @classmethod
    def to_tmux_fmt(cls) -> str:
        format_parts = ["window"]
        for field in fields(cls):
            format_parts.append(field.metadata["fmt"])
        return cls.delimeter.join(format_parts)

    @classmethod
    def from_line(cls, line: str):
        vals = [x.strip() for x in line.split(cls.delimeter)[1:]]
        obj = {}
        for index, f in enumerate(fields(cls)):
            obj[f.name] = vals[index]
        return cls(
            session_name=obj["session_name"],
            window_index=int(obj["window_index"]),
            window_name=obj["window_name"],
            window_active=bool(obj["window_active"] == "1"),
            window_flags=obj["window_flags"],
            window_layout=obj["window_layout"]
        )


@dataclass
class Pane(DataClassJsonMixin):
    delimeter: ClassVar[str] = "\t"
    session_name: str = field(default="", metadata={"fmt": "#{session_name}"})
    window_index: int = field(default=0, metadata={"fmt": "#{window_index}"})
    window_active: bool = field(default=False, metadata={
                                "fmt": "#{window_active}"})
    window_flags: str = field(default="", metadata={"fmt": ":#{window_flags}"})
    pane_index: int = field(default=0, metadata={"fmt": "#{pane_index}"})

    mytitle: str = field(default="", metadata={"fmt": "#{@mytitle}"})
    mybooter: str = field(default="", metadata={"fmt": "#{@mybooter}"})
    mypreboot: str = field(default="", metadata={"fmt": "#{@mypreboot}"})
    pane_active: bool = field(default=False, metadata={
                              "fmt": "#{pane_active}"})
    pane_current_path: str = field(
        default="", metadata={"fmt": "#{pane_current_path}"})

    @classmethod
    def to_tmux_fmt(cls) -> str:
        format_parts = ["pane"]
        for field in fields(cls):
            format_parts.append(field.metadata["fmt"])
        return cls.delimeter.join(format_parts)

    @classmethod
    def from_line(cls, line: str):
        vals = [x.strip() for x in line.split(cls.delimeter)[1:]]
        obj = {}
        for index, f in enumerate(fields(cls)):
            obj[f.name] = vals[index]
        return cls(
            session_name=obj["session_name"],
            window_index=int(obj["window_index"]),
            window_active=obj["window_active"] == "1",
            window_flags=obj["window_flags"],
            pane_index=int(obj["pane_index"]),
            mytitle=obj["mytitle"],
            mybooter=obj["mybooter"],
            mypreboot=obj["mypreboot"],
            pane_active=obj["pane_active"] == "1",
            pane_current_path=obj["pane_current_path"],
        )


class TmuxCli:
    def __init__(self):
        pass
    def lrun(self, cmd: str, **kwargs):
        print(cmd)
        if os.environ.get("TMUX_DRY_RUN", "") == "true":
            return
        return invoke.run(cmd, **kwargs)
    def run(self, cmd: str, **kwargs):
        return invoke.run(cmd, **kwargs)

    def ami_intmux(self) -> bool:
        return os.environ.get("TMUX", "") != ""

    def list_options(self, session_name: str) -> dict[str, str]:
        opts = self.run(
            f"tmux show-options  -t {session_name}", hide=True).stdout
        return {x.split()[0]: x.split()[1] for x in opts.splitlines() if x.startswith("@my_")}

    def list_pane(self, session_name: str) -> list[Pane]:
        panes = self.run(
            f"tmux list-panes -t {session_name} -F \"{Pane.to_tmux_fmt()}\"", hide=True).stdout
        ps = [Pane.from_line(x) for x in panes.splitlines()]
        ps.sort(key=lambda x: x.window_index*100+x.pane_index)
        return ps

    def cur_session(self) -> str:
        return self.run("tmux display-message -p '#S'", hide=True).stdout.strip()

    def list_window(self, session_name: str) -> list[Pane]:
        panes = self.run(
            f"tmux list-windows -t {session_name} -F \"{Window.to_tmux_fmt()}\"", hide=True).stdout
        ws = [Window.from_line(x) for x in panes.splitlines()]
        sorted(ws, key=lambda x: x.window_index)
        return ws


@dataclass
class Layout(DataClassJsonMixin):
    session_name: str
    wins: list[Window]
    panes: list[Pane]
    var: dict[str, str] = field(default_factory=dict)


class X:
    def __init__(self):
        self.cli = TmuxCli()

    def list_panel(self, name: str):
        ps = self.cli.list_pane(name)
        print(Pane.schema().dumps(ps, many=True))

    def list_window(self, name: str):
        ps = self.cli.list_window(name)
        print(Window.schema().dumps(ps, many=True))

    def _fetch_var_path(self, layout: Layout):
        env = {}
        my_var_path = layout.var.get("@my_var_path", None)
        if my_var_path is None:
            return None
        for e in my_var_path.split(","):
            env[e] = os.environ.get(e, "")
        return env

    def _resolve_var_path(self, layout: Layout):
        env = self._fetch_var_path(layout)
        if env is None:
            return
        for p in layout.panes:
            pcp = p.pane_current_path
            for k, v in env.items():
                if pcp.startswith("$"+k):
                    p.pane_current_path = pcp.replace("$"+k, v)
        pass

    def _patch_var_path(self, layout: Layout):
        env = self._fetch_var_path(layout)
        if env is None:
            return
        for p in layout.panes:
            pcp = p.pane_current_path
            for k, v in env.items():
                if pcp.startswith(v):
                    p.pane_current_path = pcp.replace(v, "$"+k)
        pass

    def save(self, output: str):
        session_name = self.cli.cur_session()
        var = self.cli.list_options(session_name)
        layout = Layout(session_name=session_name, wins=self.cli.list_window(
            session_name), panes=self.cli.list_pane(session_name), var=var)
        self._patch_var_path(layout)
        data=layout.to_json(indent=2, ensure_ascii=False)
        if output == "-":
            print(data)
            return
        pass
        p = f"{output}/{layout.session_name}.tmux.json"
        Path(p).write_text(data)
        print(p)
        pass

    def gen_tmux_send_keys(self, booter: str,without_enter:bool=False):
        if not booter:
            raise ValueError("booter is empty")

        delimeter = booter[0]
        print(f"|{booter}|   |{delimeter}|")
        cmds=[]
        if delimeter != "@":
            cmds=[booter]
            pass
        else:
            cmds = [f"'{x.strip()}'" for x in booter.removeprefix(
            delimeter).split(delimeter)]
        print(f"cmds is {cmds}")
        cmds.insert(0, "C-c")
        full = " 'enter' ".join(cmds)
        if without_enter:
            return f"{full};"
        return f"{full} 'enter';"

    def load(self, fpath: str):
        exp = Layout.from_json(Path(fpath).read_text())
        if not self.cli.ami_intmux():
            raise Exception("在tmux环境下执行")
        session_name = self.cli.cur_session()
        cur = Layout(session_name=self.cli.cur_session(), wins=self.cli.list_window(session_name),
                     panes=self.cli.list_pane(session_name))

        for k, v in exp.var.items():
            self.cli.lrun(
                f"tmux set-option -t {session_name} {k} {v}", hide=True)
        self._resolve_var_path(exp)
        for w in exp.wins:
            cur_win_index = [x.window_index for x in cur.wins]
            if w.window_index not in cur_win_index:
                self.cli.lrun(
                    f"tmux new-window -d -t \"{session_name}:{w.window_index}\"", hide=True)
            self.cli.lrun(
                f"tmux rename-window -t \"{session_name}:{w.window_index}\"  \"{w.window_name}\"", hide=True)
        # 创建panel
        for p in exp.panes:
            cur_pane_index = [x.pane_index for x in cur.panes]
            if p.pane_index not in cur_pane_index:
                self.cli.lrun(
                    f"tmux split-window -t \"{session_name}:{p.window_index}\" -c \"{p.pane_current_path}\"", hide=True)
                self.cli.lrun(
                    f"tmux select-layout -t \"{session_name}:{p.window_index}\" tiled", hide=True)
        # 设置layout
        for w in exp.wins:
            self.cli.lrun(
                f"tmux select-layout -t \"{session_name}:{w.window_index}\" \"{w.window_layout}\"", hide=True)
        # 设置option
        for p in exp.panes:
            t = f"{session_name}:{p.window_index}.{p.pane_index}"

            cmd = f"""tmux set-option -p -t "{t}" @mytitle "{p.mytitle}" """
            self.cli.lrun(cmd, hide=True)

            cmd = f"""tmux set-option -p -t "{t}" @mypreboot "{p.mypreboot}" """
            self.cli.lrun(cmd, hide=True)

            cmd = f"""tmux set-option -p -t "{t}" @mybooter "{p.mybooter}" """
            self.cli.lrun(cmd, hide=True)
        sleep(5)
        # prebooter
        print("preboot")
        for p in exp.panes:
            print(f"preboot title: {p.mytitle} index: {p.pane_index}")
            t = f"{session_name}:{p.window_index}.{p.pane_index}"
            if p.mypreboot is None or p.mypreboot.strip() == "":
                continue
            cmd = self.gen_tmux_send_keys(p.mypreboot)
            print(f"do preboot |{p.mypreboot}| cmd is {cmd}")
            if p.pane_active:
                cmd=cmd.removesuffix("'enter'")
            self.cli.lrun(f""" tmux send-keys -t {t} {cmd} """, hide=True)
        print("booter")
        sleep(3)
        for p in exp.panes:
            t = f"{session_name}:{p.window_index}.{p.pane_index}"
            if p.mybooter is None or p.mybooter.strip() == "":
                continue
            cmd = self.gen_tmux_send_keys(p.mybooter,without_enter=p.pane_active)
            self.cli.lrun(f""" tmux send-keys -t "{t}" {cmd} """, hide=True)
        pass

        # focus window and pane
        for p in exp.panes:
            if p.pane_active:
                t = f"{session_name}:{p.window_index}.{p.pane_index}"
            self.cli.lrun(
                f"""tmux switch-client -t "{session_name}:{p.window_index}" """, hide=True)
            self.cli.lrun(
                f"""tmux select-pane -t "{p.pane_index}" """, hide=True)
        for w in exp.wins:
            if w.window_active:
                self.cli.lrun(
                    f"""tmux select-window -t "{session_name}:{w.window_index}" """, hide=True)
        pass


if __name__ == "__main__":
    cmd = sys.argv[1]
    cmd = cmd.replace("-", "_").removeprefix("tmux_")
    out = X().__getattribute__(cmd)(*sys.argv[2:])
    if out:
        print(out)

File: scripts/tmux.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strings"
)

// Window represents a tmux window
type Window struct {
	SessionName  string `json:"session_name"`
	WindowIndex  int    `json:"window_index"`
	WindowName   string `json:"window_name"`
	WindowActive bool   `json:"window_active"`
	WindowFlags  string `json:"window_flags"`
	WindowLayout string `json:"window_layout"`
}

// Pane represents a tmux pane
type Pane struct {
	SessionName     string `json:"session_name"`
	WindowIndex     int    `json:"window_index"`
	WindowActive    bool   `json:"window_active"`
	WindowFlags     string `json:"window_flags"`
	PaneIndex       int    `json:"pane_index"`
	MyTitle         string `json:"mytitle"`
	MyBooter        string `json:"mybooter"`
	PaneActive      bool   `json:"pane_active"`
	PaneCurrentPath string `json:"pane_current_path"`
}

// TmuxCli represents the tmux command line interface
type TmuxCli struct{}

func (cli *TmuxCli) runCommand(command string) (string, error) {
	cmd := exec.Command("/bin/sh", "-c", command)
	output, err := cmd.CombinedOutput()
	return string(output), err
}

func (cli *TmuxCli) amiInTmux() bool {
	return os.Getenv("TMUX") != ""
}

func (cli *TmuxCli) listOptions(sessionName string) (map[string]string, error) {
	output, err := cli.runCommand(fmt.Sprintf("tmux show-options -t %s", sessionName))
	if err != nil {
		return nil, err
	}
	options := make(map[string]string)
	for _, line := range strings.Split(output, "\n") {
		if strings.HasPrefix(line, "@my_") {
			parts := strings.Fields(line)
			if len(parts) >= 2 {
				options[parts[0]] = parts[1]
			}
		}
	}
	return options, nil
}

func (cli *TmuxCli) listPane(sessionName string) ([]Pane, error) {
	// Implement the logic to list panes
	return nil, nil
}

func (cli *TmuxCli) curSession() (string, error) {
	output, err := cli.runCommand("tmux display-message -p '#S'")
	return strings.TrimSpace(output), err
}

func (cli *TmuxCli) listWindow(sessionName string) ([]Window, error) {
	// Implement the logic to list windows
	return nil, nil
}

// Layout represents a tmux layout
type Layout struct {
	SessionName string            `json:"session_name"`
	Wins        []Window          `json:"wins"`
	Panes       []Pane            `json:"panes"`
	Var         map[string]string `json:"var"`
}

// X represents the main application logic
type X struct {
	cli TmuxCli
}

func (x *X) listWindow(sessionName string) ([]Window, error) {
	_, err := x.cli.runCommand(fmt.Sprintf("tmux list-windows -t %s", sessionName))
	if err != nil {
		return nil, err
	}
	windows := []Window{}
	return windows, nil
}

func (x *X) listPane(sessionName string) ([]Pane, error) {
	out, err := x.cli.runCommand(fmt.Sprintf("tmux list-panes -t %s", sessionName))
	if err != nil {
		return nil, err
	}

	panes := []Pane{}
	return panes, nil
}

func (x *X) save() {
	sessionName, err := x.cli.curSession()
	if err != nil {
		fmt.Println("Error getting current session:", err)
		return
	}
	wins, err := x.listWindow(sessionName)
	if err != nil {
		fmt.Println("Error listing windows:", err)
		return
	}
	panes, err := x.listPane(sessionName)
	if err != nil {
		fmt.Println("Error listing panes:", err)
		return
	}
	varMap, err := x.cli.listOptions(sessionName)
	if err != nil {
		fmt.Println("Error listing options:", err)
		return
	}
	layout := Layout{
		SessionName: sessionName,
		Wins:        wins,
		Panes:       panes,
		Var:         varMap,
	}
	json.NewEncoder(os.Stdout).Encode(layout)
	os.Stdout.WriteString("\n")
}

func (x *X) genTmuxSendKeys(booter string) string {
	delimiter := strings.Split(booter, " ")[0]
	cmds := strings.Split(strings.TrimPrefix(booter, delimiter), delimiter)
	for i, cmd := range cmds {
		cmds[i] = fmt.Sprintf("'%s'", strings.TrimSpace(cmd))
	}
	cmds = append([]string{"C-c"}, cmds...)
	full := strings.Join(cmds, " 'enter' ")
	return fmt.Sprintf("%s 'enter';", full)
}

func (x *X) load(fpath string) {
	// Implement the logic to load the layout
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("No command provided")
		return
	}
	cmd := strings.Replace(os.Args[1], "-", "_", -1)
	cmd = strings.TrimPrefix(cmd, "tmux_")
	x := X{cli: TmuxCli{}}
	if os.Args[1] == "tmux-save" {
		x.save()
	} else if os.Args[1] == "tmux-load" {
		x.load(os.Args[2])
	}
}
